### PR TITLE
Support get asset id by name in pallet_assets.

### PR DESCRIPTION
**Pull request by anonymousGiga**
Support get asset id by name in pallet_assets as issue #10732 described.

### CODE REVIEW

File: frame/assets/src/impl_fungibles.rs
Comment (bkchr) [inline frame/assets/src/impl_fungibles.rs:84]:
```suggestion
			.filter_map(|v| (v.1.name == asset_name).then(|| v.0))
```

**Review by bkchr — DISMISSED** [review]

File: frame/assets/src/impl_fungibles.rs
Comment (apopiak) [inline frame/assets/src/impl_fungibles.rs:83]: Just my 2 cents: Honestly not a fan of this. `StorageMap::iter()` has unbounded weight, which makes it a footgun in on-chain development.
If you really desperately need this, I would add a separate trait with the appropriate warnings of "Unbouded execution time, DO NOT use without knowing what you're doing" etc.
(Note that none of the other traits in this file use `::iter()`.)
If you want to add this for an RPC, then I can see the motivation, but maybe would be better to add it with the RPC?

Comment (apopiak) [reply frame/assets/src/impl_fungibles.rs:83]: Also not a fan of accessing assets by name. The name can be changed any time, so it's hard to make this dependable.
I would recommend against building on-chain functionality that looks up assets by name.

**Comment by apopiak** [comment]
@bkchr Not sure I'd call this trivial.
For one it changes the fungibles API and requires downstream changes for implementers of `InspectMetadata`. Wouldn't merge this without approval from @shawntabrizi 
(Or maybe you both disagree with me.)

**Comment by bkchr** [comment]
> @bkchr Not sure I'd call this trivial. For one it changes the fungibles API and requires downstream changes for implementers of `InspectMetadata`. Wouldn't merge this without approval from @shawntabrizi (Or maybe you both disagree with me.)

1. I did not merge it.
2. I did request a review from @shawntabrizi.
3. `trivial` is about if that requires an audit and this doesn't require an audit.

File: frame/assets/src/impl_fungibles.rs
Comment (bkchr) [reply frame/assets/src/impl_fungibles.rs:83]: > Honestly not a fan of this. `StorageMap::iter()` has unbounded weight, which makes it a footgun in on-chain development.

That is a good argument! I have overseen this!

**Comment by anonymousGiga** [comment]
The actual use here is that we must use the name when using the asset, and then query the corresponding assetId. Of course, in our use case, the name is unique (it may be the address of a contract on other chains). Because there is already a name field in assetmetadata, if substrate provide a way to reverse query assetId by name, we can save some storage on the chain without store the mapping relationship between assetid and name on chain.

Other people may have similar requirements when using substrate, they may use other elements to query assetid (such as symbol, etc.).  If it is allowed to add a trait, we can query assetId through name and symbol in this trait?

**Pull request closed without merging**
